module/zfs/metaslab.c

```diff
@@ -2598,8 +2881,145 @@ metaslab_group_alloc_verify(spa_t *spa, const blkptr_t *bp, void *tag)
 #endif
 }
 
+/*
+ * block_alloc_stats() is common to metaslab layer and zdb (for auditing).
+ */
+static void
+block_alloc_stats(ms_alloc_stats_t *mas, int64_t asize, block_context_t *blkctx)
+{
+	int64_t psize;
+	int h, count;
+
+	if (asize < 0) {
+		count = -1;
+		psize = -((int64_t)blkctx->bc_psize);
+	} else {
+		count = 1;
+		psize = blkctx->bc_psize;
+	}
+
+	/*
+	 * Gang child members are not tracked
+	 */
+	if (blkctx->bc_child_gang)
+		return;
+
+	/*
+	 * transient zil blocks are never tracked
+	 */
+	if (DMU_OT_IS_ZIL(blkctx->bc_type) || blkctx->bc_type == DMU_OT_NONE)
+		return;
+
+	if (DMU_OT_IS_DDT(blkctx->bc_type)) {
+		mas->mas_dedup_bytes += psize;
+		mas->mas_dedup_blocks += count;
+		return;
+	}
+
+	/* File blocks */
+	if (DMU_OT_IS_FILE(blkctx->bc_type)) {
+		if (blkctx->bc_level == 0) {
+			mas->mas_file_bytes += psize;
+			mas->mas_file_blocks += count;
+
+			/* File small blocks (up to 128K) */
+			if (blkctx->bc_psize <= MS_SPECIAL_SMALLBLK_BOUNDS) {
+				h = highbit64(blkctx->bc_psize) -
+				    SPA_MINBLOCKSHIFT - 1;
+				if (h < CAS_HISTOGRAM_SIZE)
+					mas->mas_smallblk_histogram[h] += count;
```

Comment:
I think we should track the blocks that overflow this as well (in the highest bucket), unless you increase the histogram size to cover all possible block sizes (to SPA_MAXBLOCKSIZE).  Although I guess you could argue that we can compute the overflow based on the sum of the histogram and the mas_file_blocks.

Reply:
Updated histogram to include the larger blocks sizes